Quantized Embedding

mmoffatt2 opened the pull request:
Moved quantize embedding code from "Quantization and Binarization Implementations for Linear Layers" to new PR

gkielian commented on quantization/quantize.py:
```diff
@@ -0,0 +1,119 @@
+    "stochastic_quant": stochastic_quantize
+}
+
+_fake_quantize = FakeLinearQuantizationFunction.apply
```
Maybe remove from this CL

gkielian commented on model.py:
```diff
@@ -406,7 +416,11 @@ def update_block_size(self, new_block_size):
         if new_block_size > self.config.block_size:
             self.config.block_size = new_block_size
-            self.transformer.wpe = nn.Embedding(new_block_size, self.config.n_embd)
+            if self.config.quantize_wpe:
```
maybe add `and if self.config.use_abs_pos_embeddings`

gkielian replied:
There might be some others as well

gkielian commented on train.py:
```diff
@@ -126,6 +126,11 @@ def parse_args():
     model_group.add_argument( "--linear_std_init", type=float, default=0.02)
 
+    # Quantization Options
+    model_group.add_argument("--quantize_wte", default=None, action=argparse.BooleanOptionalAction, help="Whether the word embedding is quantized")
+    model_group.add_argument("--quantize_wpe", default=None, action=argparse.BooleanOptionalAction, help="Whether the word position embedding is quantized")
```
If you can add wte quantization precision and also one for wpe

gkielian reviewed (changes requested):
Looks good! Added feedback within single comments

mmoffatt2 commented:
Addressed PR feedback. With wte and wpe quantized, model still converges and has good inference sample output.

![image](https://github.com/user-attachments/assets/2aae5a2a-1da2-4bc3-8417-2de3250d4bfc)
![image](https://github.com/user-attachments/assets/fd75ce14-3959-401b-af9c-497fd625ba93)

gkielian commented:
Looks good! : D 

model.py One step closer to having full quantization.

Merging.
